clients/src/main/java/org/apache/kafka/clients/producer/TransactionState.java

```diff
@@ -132,4 +414,332 @@ public synchronized void incrementSequenceNumber(TopicPartition topicPartition,
+                FindCoordinatorRequest.CoordinatorType.TRANSACTION, TransactionalRequest.Priority.ADD_PARTITIONS_OR_OFFSETS, isRetry, transactionalId);
+    }
+
+    private TransactionalRequest txnOffsetCommitRequest(Map<TopicPartition, OffsetAndMetadata> offsets,
+                                                        String consumerGroupId, boolean isRetry, TransactionalRequestResult result) {
+        for (Map.Entry<TopicPartition, OffsetAndMetadata> entry : offsets.entrySet()) {
+            OffsetAndMetadata offsetAndMetadata = entry.getValue();
+            pendingTxnOffsetCommits.put(entry.getKey(),
+                    new TxnOffsetCommitRequest.CommittedOffset(offsetAndMetadata.offset(), offsetAndMetadata.metadata()));
+        }
+        return txnOffsetCommitRequest(consumerGroupId, isRetry, result);
+    }
+
+    private TransactionalRequest txnOffsetCommitRequest(String consumerGroupId, boolean isRetry, TransactionalRequestResult result) {
+        TxnOffsetCommitRequest.Builder builder = new TxnOffsetCommitRequest.Builder(consumerGroupId,
+                pidAndEpoch.producerId, pidAndEpoch.epoch, OffsetCommitRequest.DEFAULT_RETENTION_TIME, pendingTxnOffsetCommits);
+        return new TransactionalRequest(builder, new TxnOffsetCommitCallback(consumerGroupId, result),
+                FindCoordinatorRequest.CoordinatorType.GROUP, TransactionalRequest.Priority.ADD_PARTITIONS_OR_OFFSETS, isRetry, consumerGroupId);
+    }
+
+    private abstract class TransactionalRequestCallBack implements RequestCompletionHandler {
+        protected final TransactionalRequestResult result;
+
+        TransactionalRequestCallBack(TransactionalRequestResult result) {
+            this.result = result;
+        }
+
+        @Override
+        public void onComplete(ClientResponse response) {
+            if (response.requestHeader().correlationId() != inFlightRequestCorrelationId)
+                throw new IllegalStateException("Cannot have more than one transactional request in flight.");
+            resetInFlightRequestCorrelationId();
+            if (response.wasDisconnected()) {
+                reenqueue();
+            } else if (response.versionMismatch() != null) {
+                if (result != null) {
+                    result.setError(Errors.UNSUPPORTED_VERSION.exception());
+                    result.done();
+                } else {
+                    throw Errors.UNSUPPORTED_VERSION.exception();
+                }
+            } else if (response.hasResponse()) {
+                handleResponse(response.responseBody());
+            } else {
+                if (result != null) {
+                    result.setError(Errors.UNKNOWN.exception());
+                    result.done();
+                } else {
+                    throw Errors.UNKNOWN.exception();
+                }
+            }
+        }
+
+        public abstract void handleResponse(AbstractResponse responseBody);
```

Comment:
Since we're just casting the response object anyway, we could make `TransactionalRequestCallback` generic and let the type indicate the specific response type. In other words:
```java
private abstract class TransactionalRequestCallBack<T extends AbstractResponse> implements RequestCompletionHandler {
  abstract void handleResponse(T response);
}